### ISSUE TITLE
GPUI overdraws too much

### ISSUE THREAD

**Issue by kvark** [zed-industries/zed#8043]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

![zed-overdraw](https://github.com/zed-industries/zed/assets/107301/6c684f60-8de4-41ce-b9d2-e6a8e2018872)

It looks like most "ordinary" pixels are being overdraw 5-6 times on screen. See screenshot from RenderDoc's "pass overdrawn" mode. I looked at the green area (5 draws per pixel) and at the yellow one on the left (6 draws per pixel). This is burning GPU power too much unnecessarily and draining the battery.

Ideally, these areas would have 0 to 1 draw. 0 is possible because the render pass is starting with a clear color. For elements overlapping each other, there needs to be some technique for pixel rejection employed. Generally it's depth or stencil testing. For example, WebRender draws opaque elements from front to back and writes to depth, preventing deeper layers from overwriting the pixels of the front layers.

### Environment

Linux (but it unlikely matters)

**Comment by jansol** [comment]
I'm not sure GPUI has anything fully opaque? Since practically all elements are antialiased rounded rectangles which require alpha blending.

**Comment by kvark** [comment]
@jansol not all elements have rounded corners. For example, #7231 makes the fast path for such quads that are just quads, and it makes a good difference.
In WebRender we'd also split a rounded-cornered element into a "9 patch", which consists of 5 fully opaque rectangles and 4 semi-transparent rectangles. This allows drawing most of the element early as opaque.

**Comment by jansol** [comment]
Right, makes sense. And I assume it just skips elements with a non-opaque background color.

**Comment by jansol** [comment]
AFAIK there has been no change to this.

**Comment by someone13574** [comment]
> In WebRender we'd also split a rounded-cornered element into a "9 patch", which consists of 5 fully opaque rectangles and 4 semi-transparent rectangles.

Wouldn't you only need three fully opaque?

**Comment by kvark** [comment]
Yes, you can make just 3 opaque.

**Comment by jansol** [comment]
No change to this.

**Comment by jansol** [comment]
Still no changes here...

**Comment by xraxra** [comment]
this might be related to why I see GPU usage jump to ~75% usage and ~50watts here when Zed's window is visible.  I was going to post in the other [issue](https://github.com/zed-industries/zed/issues/40189)

It actually affects performance when doing gamedev, e.g. launching the build, running in a window with Zed in background or alongside, introduces sluggish framerate  (None of these issues occur with VS Code/OSS, Focus and Sublime)

i tried a couple workarounds to force it onto the integrated gpu with no luck, but launching with `VK_ICD_FILENAMES=/usr/share/vulkan/icd.d/radeon_icd.x86_64.json zed` gave a minor improvement; VRAM dropped by ~200MB consistently and GPU usage caps out at about 50%, but the high wattage is still there when the Zed window is visible, along with the reduced framerate 

Zed: 0.209.6 
OS: CachyOS 6.16.7-2
Display: Wayland (Niri 25.08 compositor)
GPUs:
NVIDIA GeForce RTX 3070 Mobile / Max-Q (discrete)
AMD Cezanne Radeon Vega (integrated)

**Comment by jansol** [comment]
> this might be related to why I see GPU usage jump to ~75% usage and ~50watts here when Zed's window is visible

Not necessarily the overdrawing so much as the fact that it redraws *everything* at the refresh rate of the display instead of limiting redraws to areas that have changed. GUI toolkits tend to have some form of damage tracking for this reason, most of the time they can just show the same image again.

**Comment by marcocondrache** [comment]
@xraxra I believe your issue is related to #29073. GPUI currently draws the entire window for one second after each input to prevent display underclocking. Since mouse movement counts as input, this results in constant GPU usage.

**Comment by SomeoneToIgnore** [comment]
As a small mitigation, https://github.com/zed-industries/zed/pull/60614 could help with `"reduce_motion": "on"` that will force all gif/spinners/etc. to not to move and avoid some part of the extra overdraws.

**Comment by DarkGhostHunter** [comment]
Any news on this?

This is probably why Zed becomes unusable after a while (had to resort to VSCode to complete a task).

**Comment by tadghh** [comment]
Would help for laptops that have high resolution displays, but cruddy integrated GPUS. Considering the whole "written in rust" point I'm curious why performance improvements like this wouldn't have been made. Anyways frame latency on the jumps to 33ms while hovering over the side panel (the hover highlight noticeably lags behind) in addition to scrolling the buffer itself. This has been present from I guess when this issues was made and up to `5a9b9558db01a6b906cec2fb70a797affdc58cdd`.

For anyone who wants a quick hack around the issue profile the `shaders.wgsl` and skip the expensive work - I genuinely cant see a difference and the frame latency issue is gone (however not an actual solution for this)
